### ISSUE TITLE
De-duplicate large sam file

### ISSUE THREAD

**Issue by morien** [samtools/samtools#2091]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib](https://github.com/samtools/htslib/releases/latest)? If not, please specify.
(run `samtools --version`)
```
samtools 1.20
Using htslib 1.20
```
 #### Please describe your environment.
- OS: `Linux 4.18.0-492.el8.x86_64`
- machine architecture: `x86_64`
- compiler: `gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-21)`

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.
I want to identify and remove "duplicate" mappings in a large (~200GB) sam file, the product of many tens of individual bam files, produced by mapping the same query file to several different reference databases (RefSeq genomes, NCBI NT database, a few custom databases of smaller size). I need to find "duplicate" mappings within this large sam file and remove them, because some reference databases may have shared content. A duplicate in this case would be the same read with the same mapping to the same reference sequence - present in multiple reference databases - irrespective of reference accession. I am not sure that duplicates of this type can be identified without merging all the bam files first.

I've been trying to figure out how to do this and I feel like I'm chasing my tail, so I've come here for help. The samtools markdup function is built for .bam files, but my .sam is too large to be accommodated in bam format (the header is too big). Picard will accept a SAM as input for its mark duplicates function, but is limited in the same way by file size constraints. I found another 3rd party tool that will mark duplicates in a sam file, but it has been 4 years since the last commit in the repo and there's a memory allocation error despite there being ~1TB of free memory on the server I'm using.

I'm here looking for alternative approaches/tricks/tips to allow de-duplication of said large sam file. Any help appreciated.

**Comment by whitwham** [comment]
**markdup** will handle sam, bam and cram files (and possibly compressed sam but I haven't tried it myself).  Most of the samtools programs will handle multiple formats.  The only ones that don't are usually there to do something file format specific.

**Comment by morien** [comment]
markdup requires input to be passed through fixmate with the -m flag, then coordinate sorted. my workflow for that would look something like this:

```
samtools fixmate -@ 10 -mu -O SAM SC-13c-TCMP_S214_L001_R1_001.fastq.merged.sam test_fixmate.sam #input for markdup must be passed through fixmate with the -m flag, then coordinate sorted
samtools sort -u -O SAM -o test_coordsort.sam -@ 10 -m 10G test_fixmate.sam	#coordinate sort
samtools markdup -@ 10 -I 300 -ru test_coordsort.sam test_rmdup.sam	#remove duplicates
```
However, I haven't been able to sort this large, uncompressed SAM (~200GB):
```
[E::bam_hdr_write] Header too long for BAM format
samtools sort: failed to write header to "test_coordsort.sam.tmp.0000.bam": No such file or directory
```
There are 439,926,763 header entries in the input file

**Comment by daviesrob** [comment]
You're right.  I thought sort managed this case by switching to SAM internally, but it actually only does so for a slightly different scenario (where a reference in longer than 2Gbases).  You could try changing [this line](https://github.com/samtools/samtools/blob/4d0d3126aac094bde4a3fd414292b1f56bf43cd5/bam_sort.c#L3488) to read:
```
                                     "wzx1", header,
```
so it always uses SAM and see if that makes it work.

As for the entire process, I'm not convinced `markdup` will do what you want.  It it sees alignments to different references it will think they're not duplicated even if the reference sequences are actually the same.  I think you'd first need to find which references are duplicates, which should be easy if you have MD5 checksums for them, then for each remove all but one from the header and repoint the alignment records that hit the removed names to use the name you kept (or possibly remove them completely, as they're likely to be the duplicates you wanted to find).  After that markdup is more likely to work, although this won't handle the cases where the references are similar but not exactly the same - those will be much harder.

The other way to do it would be to deduplicate your reference sequences at the start, and then realign to the deduplicated set.  I suspect that may be much easier and would work better.
